templates/types/streaming/fastapi/app/api/routers/chat.py

```diff
@@ -41,6 +74,13 @@ async def _chat_response_generator():
                 response = await chat_engine.astream_chat(
                     last_message_content, messages
                 )
+                source_nodes = [
+                    SourceNodes.from_source_node(node)
+                    for node in response.source_nodes
+                ]
+                process_source_nodes(source_nodes, background_tasks)
```

Comment:
We should move this to callback event handler, but we'll need a way to pass the injected `background_tasks` object to the handler.

Reply:
i think the event handler should not deal with this concern. better pass a processor function to the event handler that can deal with `source_nodes` outside of the event handler code (here downloading the llamafiles)